source/fe/fe_enriched.cc

```diff
@@ -1046,6 +1048,378 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+        -> const Function<spacedim> * {
+        AssertThrow(false,
+                    ExcMessage("Called enrichment function for FE_Nothing"));
+        return nullptr;
+      };
+
+      using EnrichmentFunctions_2DVector =
+        std::vector<std::vector<std::function<const Function<spacedim> *(
+          const typename Triangulation<dim, spacedim>::cell_iterator &)>>>;
+
+      // loop through color sets and create FE_enriched element for each
+      // of them provided before calling this function, we have color
+      // enrichment function associated with each color.
+      for (unsigned int color_set_id = 0; color_set_id != fe_sets.size();
+           ++color_set_id)
+        {
+          std::vector<const FiniteElement<dim, spacedim> *> vec_fe_enriched(
+            num_colors, &fe_nothing);
+          EnrichmentFunctions_2DVector functions(num_colors, {dummy_function});
+
+          for (auto it = fe_sets[color_set_id].begin();
+               it != fe_sets[color_set_id].end();
+               ++it)
+            {
+              // Given a color id ( = *it), corresponding color enrichment
+              // function is at index id-1 because color_enrichments is
+              // is indexed from zero.
+              const unsigned int ind = *it - 1;
+
+              AssertIndexRange(ind, vec_fe_enriched.size());
+              AssertIndexRange(ind, functions.size());
+              AssertIndexRange(ind, color_enrichments.size());
+
+              // Assume a active predicate colors {1,2} for a cell.
+              // We then need to create a vector of FE enriched elements
+              // with vec_fe_enriched[0] = vec_fe_enriched[1] = &fe_enriched
+              // which can later be associated with enrichment functions.
+              vec_fe_enriched[ind] = &fe_enriched;
+
+              // color_set_id'th color function is (color_set_id-1)
+              // element of color wise enrichments
+              functions[ind].assign(1, color_enrichments[ind]);
+            }
+
+          AssertDimension(vec_fe_enriched.size(), functions.size());
+
+          FE_Enriched<dim, spacedim> fe_component(
+            &fe_base, vec_fe_enriched, functions);
+          fe_collection->push_back(fe_component);
+        }
+    }
+  } // namespace internal
+
+
+
+  template <int dim, int spacedim>
+  Helper<dim, spacedim>::Helper(
+    const FE_Q<dim, spacedim> &                             fe_base,
+    const FE_Q<dim, spacedim> &                             fe_enriched,
+    const std::vector<predicate_function<dim, spacedim>> &  predicates,
+    const std::vector<std::shared_ptr<Function<spacedim>>> &enrichments) :
+    fe_base(fe_base),
+    fe_enriched(fe_enriched),
+    fe_nothing(1, true),
+    predicates(predicates),
+    enrichments(enrichments)
+  {}
+
+
+
+  template <int dim, int spacedim>
+  std::shared_ptr<hp::FECollection<dim, spacedim>>
+  Helper<dim, spacedim>::get_fe_collection(
+    hp::DoFHandler<dim, spacedim> &dof_handler)
```

Comment:
By the way, I think the function name is poorly chosen. A function called `get_...` should simply return something. But this function takes a non-`const` DoFHandler object, computes all sorts of things, and at the end returns an FE collection. I think naming this function better for its purpose would be useful :-)

Reply:
probably `build_fe_collection()` would better.